File: tools/validate_data.rb
#!/usr/bin/env ruby

require 'date'
require 'set'
require 'uri'
require 'yaml'

ROOT = File.expand_path('..', __dir__)

DATA_FILES = {
  '_data/current.yml' => {
    allowed: %w[name location dates url twitter status],
    required: %w[name location dates url],
    order: :ascending,
    strict: true,
    check_tracking: true
  },
  '_data/past.yml' => {
    allowed: %w[name location dates url twitter status video_playlist video_url],
    required: %w[name location dates url],
    order: nil,
    strict: false,
    check_tracking: false
  },
  '_data/closed.yml' => {
    allowed: %w[name location first_date last_date url twitter status],
    required: %w[name location first_date last_date url],
    order: nil,
    strict: true,
    check_tracking: true
  }
}.freeze

MONTH_PATTERN = '(?:January|February|March|April|May|June|July|August|September|October|November|December|Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Sept|Oct|Nov|Dec)'

def normalize_month(name)
  case name.downcase
  when 'jan' then 'January'
  when 'feb' then 'February'
  when 'mar' then 'March'
  when 'apr' then 'April'
  when 'may' then 'May'
  when 'jun' then 'June'
  when 'jul' then 'July'
  when 'aug' then 'August'
  when 'sep', 'sept' then 'September'
  when 'oct' then 'October'
  when 'nov' then 'November'
  when 'dec' then 'December'
  else
    name.capitalize
  end
end

def build_date(month_name, day, year)
  month_index = Date::MONTHNAMES.index(month_name)
  return nil unless month_index

  Date.new(year.to_i, month_index, day.to_i)
rescue ArgumentError
  nil
end

def parse_event_date(value, boundary)
  return value if value.is_a?(Date)
  return nil if value.nil?

  text = value.to_s.strip.gsub(/[–—]/, '-')

  if text =~ /^(#{MONTH_PATTERN})\s+(\d{1,2})\s*-\s*(#{MONTH_PATTERN})\s*(\d{1,2}),?\s*(\d{4})$/i
    start_month = normalize_month(Regexp.last_match(1))
    start_day = Regexp.last_match(2)
    end_month = normalize_month(Regexp.last_match(3))
    end_day = Regexp.last_match(4)
    year = Regexp.last_match(5)
  elsif text =~ /^(#{MONTH_PATTERN})\s+(\d{1,2})\s*-\s*(\d{1,2}),?\s*(\d{4})$/i
    start_month = end_month = normalize_month(Regexp.last_match(1))
    start_day = Regexp.last_match(2)
    end_day = Regexp.last_match(3)
    year = Regexp.last_match(4)
  elsif text =~ /^(#{MONTH_PATTERN})\s+(\d{1,2})\s*-\s*(\d{1,2})\s+(#{MONTH_PATTERN}),?\s*(\d{4})$/i
    start_month = normalize_month(Regexp.last_match(1))
    start_day = Regexp.last_match(2)
    end_month = normalize_month(Regexp.last_match(4))
    end_day = Regexp.last_match(3)
    year = Regexp.last_match(5)
  elsif text =~ /^(#{MONTH_PATTERN})\s+(\d{1,2}),?\s*(\d{4})$/i
    start_month = end_month = normalize_month(Regexp.last_match(1))
    start_day = end_day = Regexp.last_match(2)
    year = Regexp.last_match(3)
  end

  month = boundary == :end ? end_month : start_month
  day = boundary == :end ? end_day : start_day
  build_date(month, day, year)
rescue ArgumentError
  nil
end

def valid_url?(value)
  uri = URI.parse(value.to_s)
  uri.is_a?(URI::HTTP) && !uri.host.to_s.empty?
rescue URI::InvalidURIError
  false
end

def has_tracking_source?(value)
  uri = URI.parse(value.to_s)
  URI.decode_www_form(uri.query.to_s).any? do |key, val|
    key == 'utm_source' && val == 'testingconferences'
  end
rescue URI::InvalidURIError
  false
end

errors = []
warnings = []
seen_names = Hash.new { |hash, key| hash[key] = [] }

DATA_FILES.each do |relative_path, rules|
  path = File.join(ROOT, relative_path)
  data = YAML.safe_load(File.read(path), permitted_classes: [Date])

  unless data.is_a?(Array)
    errors << "#{relative_path}: expected a top-level array"
    next
  end

  local_names = Set.new
  previous_date = nil

  data.each_with_index do |event, index|
    label = "#{relative_path}[#{index + 1}]"

    unless event.is_a?(Hash)
      errors << "#{label}: expected a mapping"
      next
    end

    name = event['name'].to_s.strip
    display = name.empty? ? label : "#{relative_path}: #{name}"

    unknown_fields = event.keys - rules[:allowed]
    unless unknown_fields.empty?
      message = "#{display}: unknown fields: #{unknown_fields.join(', ')}"
      rules[:strict] ? errors << message : warnings << message
    end

    rules[:required].each do |field|
      value = event[field]
      next unless value.nil? || value.to_s.strip.empty?

      message = "#{display}: missing required field `#{field}`"
      rules[:strict] ? errors << message : warnings << message
    end

    if !name.empty?
      warnings << "#{display}: duplicate name in #{relative_path}" if local_names.include?(name)
      local_names.add(name)
      seen_names[name] << relative_path
    end

    if event['twitter'].to_s.include?('@')
      errors << "#{display}: twitter value should not include @"
    end

    if event.key?('url')
      url = event['url']
      errors << "#{display}: url is not a valid HTTP(S) URL" unless valid_url?(url)
      if rules[:check_tracking] && !has_tracking_source?(url)
        warnings << "#{display}: url is missing utm_source=testingconferences"
      end
    end

    if event.key?('video_playlist') && !valid_url?(event['video_playlist'])
      errors << "#{display}: video_playlist is not a valid HTTP(S) URL"
    end

    if event.key?('video_url') && !valid_url?(event['video_url'])
      warnings << "#{display}: video_url is not a valid HTTP(S) URL"
    end

    if relative_path == '_data/closed.yml'
      first_date = parse_event_date(event['first_date'], :start)
      last_date = parse_event_date(event['last_date'], :end)
      warnings << "#{display}: could not parse first_date" if event['first_date'] && first_date.nil?
      warnings << "#{display}: could not parse last_date" if event['last_date'] && last_date.nil?
      if first_date && last_date && first_date > last_date
        errors << "#{display}: first_date is after last_date"
      end
    elsif event['dates']
      order_date = parse_event_date(event['dates'], :start)
      warnings << "#{display}: could not parse dates for ordering" if order_date.nil?

      if rules[:order] && order_date && previous_date
        if rules[:order] == :ascending && order_date < previous_date
          warnings << "#{display}: appears out of chronological order"
        elsif rules[:order] == :descending && order_date > previous_date
          warnings << "#{display}: appears out of reverse chronological order"
        end
      end

      previous_date = order_date if order_date
    end
  end
rescue Psych::Exception => e
  errors << "#{relative_path}: YAML parse error: #{e.message}"
rescue Errno::ENOENT
  errors << "#{relative_path}: file not found"
end

seen_names.each do |name, paths|
  unique_paths = paths.uniq
  next if unique_paths.length == 1

  warnings << "#{name}: appears in multiple data files: #{unique_paths.join(', ')}"
end

warnings.each { |warning| warn "WARNING: #{warning}" }
errors.each { |error| warn "ERROR: #{error}" }

if errors.empty?
  puts "Conference data validation passed with #{warnings.length} warning(s)."
  exit 0
else
  warn "Conference data validation failed with #{errors.length} error(s) and #{warnings.length} warning(s)."
  exit 1
end

File: tools/validate_data.py
#!/usr/bin/env python3
import sys
import os
import re
from datetime import datetime
from urllib.parse import urlparse, parse_qs

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

DATA_FILES = {
    '_data/current.yml': {
        'allowed': {'name', 'location', 'dates', 'url', 'twitter', 'status'},
        'required': {'name', 'location', 'dates', 'url'},
        'order': 'ascending',
        'strict': True,
        'check_tracking': True
    },
    '_data/past.yml': {
        'allowed': {'name', 'location', 'dates', 'url', 'twitter', 'status', 'video_playlist', 'video_url'},
        'required': {'name', 'location', 'dates', 'url'},
        'order': None,
        'strict': False,
        'check_tracking': False
    },
    '_data/closed.yml': {
        'allowed': {'name', 'location', 'first_date', 'last_date', 'url', 'twitter', 'status'},
        'required': {'name', 'location', 'first_date', 'last_date', 'url'},
        'order': None,
        'strict': True,
        'check_tracking': True
    }
}

MONTH_MAP = {
    'jan': 1, 'january': 1,
    'feb': 2, 'february': 2,
    'mar': 3, 'march': 3,
    'apr': 4, 'april': 4,
    'may': 5,
    'jun': 6, 'june': 6,
    'jul': 7, 'july': 7,
    'aug': 8, 'august': 8,
    'sep': 9, 'sept': 9, 'september': 9,
    'oct': 10, 'october': 10,
    'nov': 11, 'november': 11,
    'dec': 12, 'december': 12
}

MONTH_PATTERN = r'(?:January|February|March|April|May|June|July|August|September|October|November|December|Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Sept|Oct|Nov|Dec)'

def parse_simple_yaml_list(filepath):
    entries = []
    curr = None
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    for line in lines:
        raw = line
        stripped = line.strip()
        if not stripped or stripped.startswith('#'):
            continue
        if stripped.startswith('- '):
            if curr is not None:
                entries.append(curr)
            curr = {}
            body = stripped[2:].strip()
            if ':' in body:
                k, v = body.split(':', 1)
                curr[k.strip()] = v.strip().strip('\"\'')
        elif curr is not None and ':' in stripped:
            k, v = stripped.split(':', 1)
            curr[k.strip()] = v.strip().strip('\"\'')
    if curr is not None:
        entries.append(curr)
    return entries

def parse_event_date(value, boundary='start'):
    if not value:
        return None
    text = str(value).strip().replace('–', '-').replace('—', '-')
    
    m = re.match(r'^(' + MONTH_PATTERN + r')\s+(\d{1,2})\s*-\s*(' + MONTH_PATTERN + r')\s*(\d{1,2}),?\s*(\d{4})$', text, re.IGNORECASE)
    if m:
        s_m, s_d, e_m, e_d, year = m.group(1), m.group(2), m.group(3), m.group(4), m.group(5)
    else:
        m = re.match(r'^(' + MONTH_PATTERN + r')\s+(\d{1,2})\s*-\s*(\d{1,2}),?\s*(\d{4})$', text, re.IGNORECASE)
        if m:
            s_m = e_m = m.group(1)
            s_d, e_d, year = m.group(2), m.group(3), m.group(4)
        else:
            m = re.match(r'^(' + MONTH_PATTERN + r')\s+(\d{1,2}),?\s*(\d{4})$', text, re.IGNORECASE)
            if m:
                s_m = e_m = m.group(1)
                s_d = e_d = m.group(2)
                year = m.group(3)
            else:
                return None

    month_name = e_m if boundary == 'end' else s_m
    day = e_d if boundary == 'end' else s_d
    m_idx = MONTH_MAP.get(month_name.lower())
    if not m_idx:
        return None
    try:
        return datetime(int(year), m_idx, int(day))
    except ValueError:
        return None

def valid_url(url_str):
    try:
        u = urlparse(url_str)
        return u.scheme in ('http', 'https') and bool(u.netloc)
    except Exception:
        return False

def has_tracking_source(url_str):
    try:
        u = urlparse(url_str)
        qs = parse_qs(u.query)
        return 'testingconferences' in qs.get('utm_source', [])
    except Exception:
        return False

def main():
    errors = []
    warnings = []
    seen_names = {}

    for rel_path, rules in DATA_FILES.items():
        abs_path = os.path.join(ROOT, rel_path)
        if not os.path.exists(abs_path):
            errors.append(f"{rel_path}: file not found")
            continue
        
        data = parse_simple_yaml_list(abs_path)
        local_names = set()
        previous_date = None

        for index, event in enumerate(data, 1):
            label = f"{rel_path}[{index}]"
            name = event.get('name', '').strip()
            display = f"{rel_path}: {name}" if name else label

            unknown_fields = set(event.keys()) - rules['allowed']
            if unknown_fields:
                msg = f"{display}: unknown fields: {', '.join(sorted(unknown_fields))}"
                if rules['strict']:
                    errors.append(msg)
                else:
                    warnings.append(msg)

            for req in rules['required']:
                if not event.get(req):
                    msg = f"{display}: missing required field `{req}`"
                    if rules['strict']:
                        errors.append(msg)
                    else:
                        warnings.append(msg)

            if name:
                if name in local_names:
                    warnings.append(f"{display}: duplicate name in {rel_path}")
                local_names.add(name)
                seen_names.setdefault(name, []).append(rel_path)

            if '@' in event.get('twitter', ''):
                errors.append(f"{display}: twitter value should not include @")

            if 'url' in event:
                u = event['url']
                if not valid_url(u):
                    errors.append(f"{display}: url is not a valid HTTP(S) URL")
                elif rules['check_tracking'] and not has_tracking_source(u):
                    warnings.append(f"{display}: url is missing utm_source=testingconferences")

            if 'video_playlist' in event and not valid_url(event['video_playlist']):
                errors.append(f"{display}: video_playlist is not a valid HTTP(S) URL")

            if 'video_url' in event and not valid_url(event['video_url']):
                warnings.append(f"{display}: video_url is not a valid HTTP(S) URL")

            if rel_path == '_data/closed.yml':
                f_date = parse_event_date(event.get('first_date'), 'start')
                l_date = parse_event_date(event.get('last_date'), 'end')
                if event.get('first_date') and not f_date:
                    warnings.append(f"{display}: could not parse first_date")
                if event.get('last_date') and not l_date:
                    warnings.append(f"{display}: could not parse last_date")
                if f_date and l_date and f_date > l_date:
                    errors.append(f"{display}: first_date is after last_date")
            elif event.get('dates'):
                order_date = parse_event_date(event.get('dates'), 'start')
                if not order_date and 'TBA' not in event.get('dates', '') and not re.search(r'20\d\d', event.get('dates','')):
                    warnings.append(f"{display}: could not parse dates for ordering")

                if rules['order'] and order_date and previous_date:
                    if rules['order'] == 'ascending' and order_date < previous_date:
                        warnings.append(f"{display}: appears out of chronological order ({order_date.strftime('%Y-%m-%d')} < {previous_date.strftime('%Y-%m-%d')})")
                if order_date:
                    previous_date = order_date

    for w in warnings:
        print(f"WARNING: {w}", file=sys.stderr)
    for e in errors:
        print(f"ERROR: {e}", file=sys.stderr)

    if not errors:
        print(f"Conference data validation passed with {len(warnings)} warning(s).")
        sys.exit(0)
    else:
        print(f"Conference data validation failed with {len(errors)} error(s) and {len(warnings)} warning(s).", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()
